### PR TITLE
many: refactor to kernel/fde and allow `fde-setup initial-setup` to return json

### PR DESCRIPTION
**Pull request by mvo5**
This commit allows the fde-setup hook to return json instead of
raw binary data. This is broken out from PR #10149 for easier
review.

This also now contains a refactoring starting to move many definitions and details about FDE kernel hooks to kernel/fde.

### CODE REVIEW

File: boot/seal.go [hunk 1]
```diff
@@ -141,20 +139,43 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
...
+	EncryptedKey []byte `json:"encrypted-key"`
+	Handle       []byte `json:"handle"`
+}
+
 func sealKeyToModeenvUsingFDESetupHook(key, saveKey secboot.EncryptionKey, model *asserts.Model, modeenv *Modeenv) error {
 	// TODO: support full boot chains
 
 	for _, skr := range append(runKeySealRequests(key), fallbackKeySealRequests(key, saveKey)...) {
+		var unencryptedPayload [len(skr.Key)]byte
+		copy(unencryptedPayload[:], skr.Key[:])
```
Comment (bboozzoo) [inline boot/seal.go:152]: Do we need a copy here? AFAIU DevieManger.runFDESetupHook will not modify this, neither should the mocking in tests.

File: boot/seal.go [hunk 2]
```diff
@@ -141,20 +139,43 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
 	}
 }
 
+type fdeSetupHookResult struct {
+	EncryptedKey []byte `json:"encrypted-key"`
+	Handle       []byte `json:"handle"`
```
Comment (bboozzoo) [inline boot/seal.go:144]: How is handle used later on?

File: boot/seal.go [hunk 3]
```diff
@@ -141,20 +139,43 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
 	}
 }
 
+type fdeSetupHookResult struct {
```
Comment (bboozzoo) [inline boot/seal.go:142]: Maybe we should include v2 in the name?

**Review by bboozzoo — APPROVED** [review]
LGTM

File: boot/seal.go (hunk 2 above)
Comment (pedronis) [reply boot/seal.go:144]: it's meant to be passed back when we unseal the key

File: boot/seal.go [hunk 4]
```diff
@@ -141,20 +139,43 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
...
+		// hook at this point. However the "denver" project
+		// uses the old and deprecated v1 API that returns raw
+		// bytes and we still need to support this.
+		var res fdeSetupHookResult
+		if err := json.Unmarshal(hookOutput, &res); err != nil {
+			// If the input is not json and matches the
+			// size of the input encrypton key we assume
+			// we deal with a v1 hook that passes us back
+			// raw binary data instead of json.
+			if len(hookOutput) != len(unencryptedPayload) {
```
Comment (pedronis) [inline boot/seal.go:171]: generally hookOutput will be likely larger than the input key so this check is not meaningful

**Review by pedronis — COMMENTED** [review]
two comments

File: boot/seal.go [hunk 5]
```diff
@@ -71,11 +71,9 @@ const (
 
 // FDESetupHookParams contains the inputs for the fde-setup hook
 type FDESetupHookParams struct {
-	Key     secboot.EncryptionKey
+	Key     []byte
 	KeyName string
```
Comment (pedronis) [inline boot/seal.go:75]: we can drop this afaict?

File: boot/seal.go (hunk 2 above)
Comment (pedronis) [reply boot/seal.go:144]: afaict from the intention of secboot code, here we should probably use *json.RawMessage for Handle and not []byte, of course the hooks can send a base64 string and that's fine

File: boot/seal.go [hunk 6]
```diff
@@ -141,20 +139,43 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
 	}
 }
 
+type fdeSetupHookResult struct {
+	EncryptedKey []byte `json:"encrypted-key"`
```
Comment (pedronis) [inline boot/seal.go:143]: if we must keep sealed-key on reaveal, this name is strange.  We need to discuss what we can or can't change at this point.

File: boot/seal.go [hunk 7]
```diff
@@ -141,20 +139,43 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
...
+		if err := json.Unmarshal(hookOutput, &res); err != nil {
+			// If the input is not json and matches the
+			// size of the input encrypton key we assume
+			// we deal with a v1 hook that passes us back
+			// raw binary data instead of json.
+			if len(hookOutput) != len(unencryptedPayload) {
+				return fmt.Errorf("cannot decode hook output for key %s %q: %v", skr.KeyFile, hookOutput, err)
+			}
+			// v1 hooks do not support a handle
+			res.Handle = nil
```
Comment (pedronis) [inline boot/seal.go:175]: I think we want to set a special value for handle for the v1 case, not just nil. Unless we prohibit for v2 hooks not to set it at all, which we don't quite check I think

File: boot/seal.go [hunk 8]
```diff
@@ -141,20 +139,43 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
...
 		if err != nil {
 			return err
 		}
-		if err := osutil.AtomicWriteFile(filepath.Join(skr.KeyFile), sealedKey, 0600, 0); err != nil {
+		// We expect json output that fits fdeSetupHookResult
+		// hook at this point. However the "denver" project
+		// uses the old and deprecated v1 API that returns raw
+		// bytes and we still need to support this.
+		var res fdeSetupHookResult
+		if err := json.Unmarshal(hookOutput, &res); err != nil {
```
Comment (pedronis) [inline boot/seal.go:166]: it's a bit unclear if this code belongs here or inside RunFDESetupHook. The fact that we didn't get to reorg secboot vs boot vs devicestate is biting us, it's very hard to follow the code spread over so many places :/

File: boot/seal.go [hunk 9]
```diff
@@ -141,20 +139,43 @@ func fallbackKeySealRequests(key, saveKey secboot.EncryptionKey) []secboot.SealK
 	}
 }
 
+type fdeSetupHookResult struct {
+	EncryptedKey []byte `json:"encrypted-key"`
+	Handle       []byte `json:"handle"`
+}
+
 func sealKeyToModeenvUsingFDESetupHook(key, saveKey secboot.EncryptionKey, model *asserts.Model, modeenv *Modeenv) error {
```
Comment (pedronis) [inline boot/seal.go:147]: should we pass a different value to stampSealedKeys for v2 hooks? or is not worth it?

**Review by pedronis — COMMENTED** [review]
a bunch of other considerations

File: boot/seal.go (hunk 9 above)
Comment (mvo5) [reply boot/seal.go:147]: So far the code tries to keep all filenames the same and do pure content detection. So I'm inclined to keep it as is or switch away from pure content detection.

File: boot/seal.go (hunk 4 above)
Comment (mvo5) [reply boot/seal.go:171]: Yeah, I moved this to the prefix detection instead of size based.

File: boot/seal.go (hunk 8 above)
Comment (mvo5) [reply boot/seal.go:166]: I was on the fence too, happy to move it if you think that's clearer

File: boot/seal.go (hunk 7 above)
Comment (mvo5) [reply boot/seal.go:175]: sure, happy to set it to something special, anything in particular you have in mind? `v1-hook` maybe?

File: boot/seal.go (hunk 3 above)
Comment (mvo5) [reply boot/seal.go:143]: I was thinking about this but would love to have as few references to v2 if possible, v2 is the "normal" case (until v3 ;)

File: boot/seal.go (hunk 7 above)
Comment (pedronis) [reply boot/seal.go:178]: "{v1-no-handle: true}" perhaps ?

File: boot/seal.go (hunk 3 above)
Comment (pedronis) [reply boot/seal.go:143]: this actually should probably live in devicestate/fde (which in itself maybe should be moved, but that can happen later) ?

File: boot/seal.go (hunk 8 above)
Comment (pedronis) [reply boot/seal.go:172]: I think we should move it, but really RunFDESetupHook should probably use a helper itself, some kind of UnmarshalFDESetupResult also in devicestate/fde ?

**Review by pedronis — COMMENTED** [review]
couple more considerations

File: boot/seal.go (hunk 3 above)
Comment (mvo5) [reply boot/seal.go]: I moved that over now

Comment (mvo5) [reply boot/seal.go]: Well, I move it next to FDESetupHookParams because right now this is used in "boot" and I don't think we want to import "overlord/devicemanager/fde" from there - however we could move the entire devicemanager/fde into boot/fde or secboot/fde - I don't think we do anything overlord-ish there. Happy to do this here or in a followup if desired.

File: boot/seal.go (hunk 6 above)
Comment (mvo5) [reply boot/seal.go]: I reached out to ondra about this.

File: boot/seal.go [hunk 10]
```diff
@@ -74,12 +74,17 @@ type FDESetupHookParams struct {
...
-	Models []*asserts.Model
-
 	//TODO:UC20: provide bootchains and a way to track measured
 	//boot-assets
 }
 
+// FdeSetupHookResult contains the outputs of the fde-setup hook
+type FDESetupHookResult struct {
+	// result when called with "initial-setup"
+	EncryptedKey []byte           `json:"encrypted-key"`
```
Comment (mvo5) [inline boot/seal.go:84]: @pedronis I just talked to Ondra and it's still possible to update their v2 hook - happy to change this to a better name.

Comment (pedronis) [reply boot/seal.go:84]: I think this needs to be sealed-key (and match Ondra's code), until we know what we can/can't do about the reveal side

**Comment by pedronis** [comment]
I have worked on https://github.com/pedronis/snappy/commit/887e0844b0d216dd2f1176a5f4ee492cdd0dfd88 to try to clean up some the debt we have already in this area, especially definition and details of fde-setup split between in boot or known to devicestate instead of centralized.

**Comment by pedronis** [comment]
@mvo5 @bboozzoo I pushed my refactoring

**Review by pedronis — APPROVED** [review]
this has my +1 now. It needs re-viewew by @bboozzoo  and a +1 from @mvo5

**Review by bboozzoo — APPROVED** [review]
LGTM with the last commit from @pedronis

**Review by mvo5 — COMMENTED** [review]
Thank you, :+1: from me as well
